File: 1026.maximum-difference-between-node-and-ancestor.cs
/*
 * @lc app=leetcode id=1026 lang=csharp
 *
 * [1026] Maximum Difference Between Node and Ancestor
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution 
{
    int ans = 0;
    public int[] dfs(TreeNode root)
    {
        int[] result = new int[2] {root.val, root.val};
        if (root.left != null)
        {
            int[] resultLeft = dfs(root.left);
            ans = Math.Max(ans, Math.Abs(root.val - resultLeft[0]));
            ans = Math.Max(ans, Math.Abs(root.val - resultLeft[1]));
            result[0] = Math.Min(result[0], resultLeft[0]);
            result[1] = Math.Max(result[1], resultLeft[1]);
        }
        if (root.right != null)
        {
            int[] resultRight = dfs(root.right);
            ans = Math.Max(ans, Math.Abs(root.val - resultRight[0]));
            ans = Math.Max(ans, Math.Abs(root.val - resultRight[1]));
            result[0] = Math.Min(result[0], resultRight[0]);
            result[1] = Math.Max(result[1], resultRight[1]);
        }
        return result;
    }
    public int MaxAncestorDiff(TreeNode root) 
    {
        int[] resultRoot = dfs(root);
        return ans;
    }
}
// @lc code=end

File: 1026.maximum-difference-between-node-and-ancestor.py
#
# @lc app=leetcode id=1026 lang=python3
#
# [1026] Maximum Difference Between Node and Ancestor
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # We are looking for absolute differences between a node's val and all of its descendents. The parents will be taken care of by a recursive function (dfs)
    # To simplify the process and get the absolute difference between a node and all its offspring, for example between a and {b1, b2, b3...}
    # We can sort bns. Assume smallest is b1 and largest is bn
    # When we calculate the maximum difference, the answer is max(abs(a-b1), abs(a-bn))
    # As for the recursive nature, we return and max and min of {a, b1,..., bn} for a's parent
    
    def dfs(self, root):
        result = [root.val, root.val]
        if root.left:
            left_min, left_max = self.dfs(root.left)
            self.ans = max(self.ans, abs(root.val - left_min), abs(root.val - left_max))
            result = min(result[0], left_min), max(result[1], left_max)
        if root.right:
            right_min, right_max = self.dfs(root.right)
            self.ans = max(self.ans, abs(root.val - right_min), abs(root.val - right_max))
            result = min(result[0], right_min), max(result[1], right_max)
        return result
    
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        self.ans = 0
        result = self.dfs(root)
        return self.ans
# @lc code=end

